kornia/feature/keynet.py

```diff
@@ -0,0 +1,315 @@
+        # Load extraction configuration
+        self.num_pyramid_levels: int = keynet_conf['pyramid_levels']
+        self.num_upscale_levels: int = keynet_conf['up_levels']
+        self.scale_factor_levels: float = keynet_conf['scale_factor_levels']
+        self.mr_size: float = keynet_conf['s_mult']
+        self.nms_size: int = keynet_conf['nms_size']
+        self.nms = NonMaximaSuppression2d((self.nms_size, self.nms_size))
+        self.num_features = num_features
+        self.ori = ori_module
+        self.aff = aff_module
+
+    def remove_borders(self, score_map, borders: int = 15):
+        """It removes the borders of the image to avoid detections on the corners."""
+        shape = score_map.shape
+        mask = torch.ones_like(score_map)
+
+        mask[:, :, 0:borders, :] = 0
+        mask[:, :, :, 0:borders] = 0
+        mask[:, :, shape[2] - borders:shape[2], :] = 0
+        mask[:, :, :, shape[3] - borders:shape[3]] = 0
+        return mask * score_map
+
+    def detect_features_on_single_level(self, level_img: torch.Tensor, num_kp: int, factor: Tuple[float, float]):
+        det_map = self.nms(self.remove_borders(self.model(level_img)))
+        device = level_img.device
+        dtype = level_img.dtype
+        yx = det_map.nonzero()[:, 2:].t()
+        scores = det_map[0, 0, yx[0], yx[1]]  # keynet supports only non-batched images
+
+        scores_sorted, indices = torch.sort(scores, descending=True)
+
+        indices = indices[torch.where(scores_sorted > 0.)]
+        yx = yx[:, indices[:num_kp]].t()
+        current_kp_num = len(yx)
+        xy_projected = yx.view(1, current_kp_num, 2).flip(2) * torch.tensor(factor, device=device, dtype=dtype)
+        scale_factor = 0.5 * (factor[0] + factor[1])
+        scale = scale_factor * self.mr_size * torch.ones(1, current_kp_num, 1, 1, device=device, dtype=dtype)
+        lafs = laf_from_center_scale_ori(xy_projected,
+                                         scale,
+                                         torch.zeros(1, current_kp_num, 1, device=device, dtype=dtype))
+        return scores_sorted[:num_kp], lafs
+
+    def detect(  # type: ignore
+        self, img: torch.Tensor, mask: Optional[torch.Tensor] = None  # type: ignore
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Compute points per level
+        num_features_per_level: List[float] = []
+        tmp = 0.0
+        factor_points = (self.scale_factor_levels ** 2)
+        levels = self.num_pyramid_levels + self.num_upscale_levels + 1
+        for idx_level in range(levels):
+            tmp += factor_points ** (-1 * (idx_level - self.num_upscale_levels))
+            nf = self.num_features * factor_points ** (-1 * (idx_level - self.num_upscale_levels))
+            num_features_per_level.append(nf)
+        num_features_per_level = list(map(lambda x: int(x / tmp), num_features_per_level))
+
+        _, _, h, w = img.shape
+        img_up = img.clone()
+        cur_img = img.clone()
```

Comment:
why do we need clone ?

Reply:
because we do both up-sampling and downsampling.